pkg/crawler/crawler.go

```diff
@@ -262,3 +313,260 @@ func (c *Crawler) fetchSHA1(url string) ([]byte, error) {
+	if len(errs) > 0 {
+		log.Println("errors in license classification ", errs)
+	}
+
+	// extract results
+	results := c.classifier.GetResults()
+	sort.Sort(results)
+
+	// process results to update the normalizedLicenseMap
+	if results.Len() > 0 {
+		for _, r := range results {
+			if licenseVal, ok := filesLicenseMap[r.Filename]; ok {
+				// since results are sorted, we can skip processing of data with confidence <90%
+				if r.Confidence < 0.9 {
+					break
+				}
+
+				// skip processing since a higher confidence result is already processed
+				if licenseVal.ClassificationConfidence > r.Confidence {
+					continue
+				}
+
+				licenseVal.ClassificationConfidence = r.Confidence
+				filesLicenseMap[r.Filename] = licenseVal
+
+				// update normalized license map
+				normalizedLicenseMap[licenseVal.LicenseKey] = r.Name
+			}
+		}
+	}
+
+	defer func() {
+		// update normalized license map for license keys which couldn't be classified or had no url in pom for classification
+		uniqLicenseKeys := c.uniqueLicenseKeys.Items()
+		for key, license := range uniqLicenseKeys {
+			if _, ok := normalizedLicenseMap[key]; !ok {
+				if len(license.Name) > 0 {
+					normalizedLicenseMap[key] = license.Name
+				}
+			}
+		}
+
+		err := fileutil.WriteJSON(c.licensedir+types.NormalizedlicenseFileName, normalizedLicenseMap)
+		if err != nil {
+			log.Println(err)
+		}
+	}()
+
+	return nil
+}
+
+func (c *Crawler) prepareClassifierData(ctx context.Context) ([]licenseFilesMeta, error) {
+	log.Println("Preparing license classifier data")
+
+	var licenseFiles []licenseFilesMeta
+
+	// switch from concurrent to normal map
+	uniqLicenseKeyMap := c.uniqueLicenseKeys.Items()
+	uniqueLicenseKeyList := c.uniqueLicenseKeys.Keys()
+
+	client := http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	licenseKeyChannel := make(chan string, len(uniqueLicenseKeyList))
+
+	log.Printf("Total license keys to be processed %d", len(uniqueLicenseKeyList))
+
+	// dump license keys to the channel so that they can be processed
+	for _, key := range uniqueLicenseKeyList {
+		licenseKeyChannel <- key
+	}
+
+	limit := semaphore.NewWeighted(c.opt.Limit)
+
+	// error channel
+	errCh := make(chan error)
+	defer close(errCh)
+
+	// status channel to track processing of license keys
+	type status struct {
+		Meta licenseFilesMeta
+		Done bool
+	}
+	prepStatus := make(chan status, len(uniqueLicenseKeyList))
+	defer close(prepStatus)
+
+	// process license keys channel
+	go func() {
+		for licenseKey := range licenseKeyChannel {
+
+			if err := limit.Acquire(ctx, 1); err != nil {
+				errCh <- xerrors.Errorf("semaphore acquire error: %w", err)
+			}
+
+			// process license key to generate license file
+			go func(licenseKey string) {
+				defer limit.Release(1)
+
+				licenseFileName := getLicenseFileName(c.licensedir, licenseKey)
+				licenseMeta := uniqLicenseKeyMap[licenseKey]
+				ok, err := c.generateLicenseFile(client, licenseFileName, licenseMeta)
```

Comment:
Why do we need to generate license files? It accepts `[]byte` or `io.Reader`.
https://github.com/google/licenseclassifier/blob/c1ed8fcf4babdf7c37d872cf6da5b1c32907d34c/v2/classifier.go#L321

Reply:
The v2 license classifier tool loads assets i.e gathers locally available license data https://github.com/google/licenseclassifier/blob/main/v2/tools/identify_license/backend/backend.go#L50 which is used to find appropriate matches for supplied input.

The code you pointed out would require us providing a local directory all assets if I am not wrong. It is doable but we'll have to maintain a copy locally I believe.

Reply:
>The code you pointed out would require us providing a local directory with all assets/training data if I am not wrong.

No, we don't have to do that.

Reply:
Didn't get you.

https://github.com/google/licenseclassifier/blob/main/v2/classifier.go#L278
https://github.com/google/licenseclassifier/blob/main/v2/classifier_test.go#L44

Am I missing something?